shared/actions/devices.js

```diff
@@ -1,138 +1,192 @@
+function * _deviceRemoveSaga (removeAction: RemoveDevice): SagaGenerator<any, any> {
+  // Revoking the current device uses the "deprovision" RPC instead.
+  const {currentDevice, name, deviceID} = removeAction.payload
+  if (currentDevice) {
+    try {
+      const username = yield select(state => state.config && state.config.username)
       if (!username) {
-        console.warn('No username in removeDevice')
-        return
+        const error = {errorText: 'No username in removeDevice'}
+        console.warn(error)
+        yield put(({
+          type: Constants.deviceRemoved,
+          payload: error,
+          error: true,
+        }: DeviceRemoved))
       }
-      loginDeprovisionRpc({
-        param: {username, doRevoke: true},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
-
-          if (!error) {
-            dispatch(loadDevices())
-            dispatch(setRevokedSelf(name))
-            dispatch(navigateTo([], loginTab))
-            dispatch(switchTab(loginTab))
-          }
-        },
-      })
-    } else {
-      revokeRevokeDeviceRpc({
+      yield call(loginDeprovisionRpcPromise, {param: {username, doRevoke: true}})
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
+      }: DeviceRemoved))
+      yield put(loadDevices())
+      yield put(setRevokedSelf(name))
+      yield put(navigateTo([], loginTab))
+      yield put(switchTab(loginTab))
+    } catch (e) {
+      console.warn('Error removing the current device:', e)
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: e,
+        error: true,
+      }: DeviceRemoved))
+    }
+  } else {
+    // Not the current device.
+    try {
+      yield call(revokeRevokeDeviceRpcPromise, {
         param: {deviceID, force: false},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
+      })
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
+      }: DeviceRemoved))
+    } catch (e) {
+      console.warn('Error removing a device:', e)
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: e,
+        error: true,
+      }: DeviceRemoved))
+    }
+  }
+  yield put(loadDevices())
+  yield put(navigateUp(devicesTab, Map({path: 'root'})))
```

Comment:
not yet, but we used to have a really complicated action that did something similar. We took it out because it was too complicated. This sounds like a simple solution. We haven't done it in other places because we just started using sagas